### DISCUSSION TITLE
Porting the system to the ARM architecture

### DISCUSSION THREAD

**Discussion by LukovDev** [CosmosOS/Cosmos#2416] [General]
Will Cosmos be able to build under ARM in the future?

**Comment by AsertCreator** [comment]
possibly will. but not in next 1-2 years. cosmos is still isnt very stable, it filled with bugs and still runs on x86-32 arch. it doesnt even have multithreading, apic, full acpi and other features. il2cpu (cosmos compiler) is built from the ground up with x86 in mind. adapting it for arm is very hard. adapting cosmos is another problem, cosmos is using a lot of x86 cpu apis.
these are not compatible:
- console
- graphics
- pci control
- timers
- memory management
- interrupt management
as you can see, there are no cosmos crucial components compatible with arm

arm is that thing where you need to implement everything yourself

**Reply by AnErrupTion** [reply]
> arm is that thing where you need to implement everything yourself

Not really. It's just that getting the addresses of e.g. the PCI controller is really time-consuming because it's different on each ARM platform, since anyone can make an ARM CPU (unlike x86).

**Reply by AsertCreator** [reply]
oh. im dont know **_that_** much about arm, but i clearly know that arm implementations are very confusing and they always different, porting operating system to it is very hard, because almost everything is hidden all over the memory or i/o ports
